src/rest.cpp

```diff
@@ -267,13 +262,57 @@ int Rest::Transpose(FunctorParams *functorParams)
     TransposeParams *params = dynamic_cast<TransposeParams *>(functorParams);
     assert(params);
 
-    // if (!this->HasOloc() || !this->HasPloc()) return FUNCTOR_SIBLINGS;
+    if ((!HasOloc() || !HasPloc()) && !HasLoc()) return FUNCTOR_SIBLINGS;
+
+    // Find whether current layer is top, middle (either one if multiple) or bttom
+    Staff *parentStaff = dynamic_cast<Staff *>(GetFirstAncestor(STAFF));
+    assert(parentStaff);
 
-    LogDebug("relocating rest");
+    Layer *parentLayer = dynamic_cast<Layer *>(GetFirstAncestor(LAYER));
+    assert(parentLayer);
 
-    TransPitch restLoc = this->GetTransLoc();
-    params->m_transposer->Transpose(restLoc);
-    this->UpdateFromTransLoc(restLoc);
+    Layer *firstLayer = dynamic_cast<Layer *>(parentStaff->GetFirst());
+    Layer *lastLayer = dynamic_cast<Layer *>(parentStaff->GetLast());
+
+    const bool isTopLayer = (firstLayer->GetN() == parentLayer->GetN());
+    const bool isBottomLayer = (lastLayer->GetN() == parentLayer->GetN());
+
```

Comment:
A general comment about the following if/else code is that music in percussion clefs should not be transposed.  This is because the lines/spaces on the percussion staff do not represent pitches, but rather separate instruments (as in a drum kit) or different techniques on the same instrument.  So both pitches and rests should not be transposed when the music is in a percussion clef.    This would perhaps be outside the scope of this pull request, but somewhere in the code, any request to transpose music on percussion clef should be ignored.  @lpugin can comment on where/when transposition suppression for percussion clefs should be done.

Example:

<img width="440" alt="Screen Shot 2020-08-11 at 9 34 15 AM" src="https://user-images.githubusercontent.com/3487289/89923943-233db980-dbb6-11ea-9863-a0589c3499e9.png">


```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2020-08-11T09:29:33" version="3.0.0-dev-38985cb">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000000571948533">
    <score xml:id="score-0000001669376790">
     <scoreDef xml:id="scoredef-0000000008476070">
      <staffGrp xml:id="staffgrp-0000001777087829" symbol="brace" bar.thru="true">
       <staffDef xml:id="staffdef-0000000435663544" n="1" lines="5">
        <clef xml:id="clef-L3F2" shape="G" line="2" />
       </staffDef>
       <staffDef xml:id="staffdef-0000000344893152" n="2" lines="1">
        <clef xml:id="clef-L3F1" shape="perc" line="1" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-0000001324141884" n="1">
        <layer xml:id="layer-L1F2N1" n="1">
         <note xml:id="note-L5F2" dur="2" oct="4" pname="g" accid.ges="n" />
         <note xml:id="note-L7F2" dur="2" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-0000000103187005" n="2">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="4"  loc="0" />
         <rest xml:id="rest-L6F1" dur="4" />
         <note xml:id="note-L7F1" dur="4"  loc="0" />
         <rest xml:id="rest-L8F1" dur="4" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

Currently verovio is transposing this correctly, but perhaps by accident.  Here is transposition up a major second:

```bash
verovio --transpose M2 file.mei
```

<img width="444" alt="Screen Shot 2020-08-11 at 9 38 21 AM" src="https://user-images.githubusercontent.com/3487289/89924127-6c8e0900-dbb6-11ea-8d6f-c3c5a9701142.png">

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2020-08-11T09:29:33" version="3.0.0-dev-38985cb">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000000571948533">
    <score xml:id="score-0000001669376790">
     <scoreDef xml:id="scoredef-0000000008476070">
      <staffGrp xml:id="staffgrp-0000001777087829" symbol="brace" bar.thru="true">
       <staffDef xml:id="staffdef-0000000435663544" n="1" lines="5">
        <clef xml:id="clef-L3F2" shape="G" line="2" />
       </staffDef>
       <staffDef xml:id="staffdef-0000000344893152" n="2" lines="1">
        <clef xml:id="clef-L3F1" shape="perc" line="1" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-0000001324141884" n="1">
        <layer xml:id="layer-L1F2N1" n="1">
         <note xml:id="note-L5F2" dur="2" oct="4" pname="a" accid.ges="n" />
         <note xml:id="note-L7F2" dur="2" oct="4" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-0000000103187005" n="2">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="4" loc="0" />
         <rest xml:id="rest-L6F1" dur="4" />
         <note xml:id="note-L7F1" dur="4" loc="0" />
         <rest xml:id="rest-L8F1" dur="4" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

Reply:
From what I see, this happen due to the fact that Verovio doesn't exactly transpose rests, unless they are part of the beam. Your example MEI file will be transposed the same way as it's right now - notice that there are no loc/ploc/oloc attributes for the rests on the percussion staff. I'm not sure if there's a point in having those attributes for such rests, but a condition to ignore such staff during transposition can be added if it's point of concern.